contracts/libraries/Safe150Migration.sol

```diff
@@ -0,0 +1,186 @@
+
+    // Address of Safe contract version 1.5.0 Singleton (L1)
+    // TODO: Update this address when the Safe 1.5.0 Singleton is deployed
+    address public constant SAFE_150_SINGLETON = address(0x88627c8904eCd9DF96A572Ef32A7ff13b199Ed8D);
+
+    // Address of Safe contract version 1.5.0 Singleton (L2)
+    // TODO: Update this address when the Safe 1.5.0 Singleton (L2) is deployed
+    address public constant SAFE_150_SINGLETON_L2 = address(0x0Ee37514644683f7EB9745a5726C722DeBa77e52);
+
+    // Address of Safe contract version 1.5.0 Compatibility Fallback Handler
+    // TODO: Update this address when the Safe 1.5.0 Compatibility Fallback Handler is deployed
+    address public constant SAFE_150_FALLBACK_HANDLER = address(0x8aa755cB169991fEDC3E306751dCb71964A041c7);
+
+    bytes32 internal constant GUARD_STORAGE_SLOT = 0x4a204f620c8c5ccdca3fd54d003badd85ba500436a431f0cbda4f558c93c34c8;
+
+    /**
+     * @notice Constructor
+     * @dev Initializes the migrationSingleton with the contract's own address.
+     */
+    constructor() {
+        MIGRATION_SINGLETON = address(this);
+
+        require(isContract(SAFE_150_SINGLETON), "Safe 1.4.1 Singleton is not deployed");
+        require(isContract(SAFE_150_SINGLETON_L2), "Safe 1.4.1 Singleton (L2) is not deployed");
+        require(isContract(SAFE_150_FALLBACK_HANDLER), "Safe 1.4.1 Fallback Handler is not deployed");
+    }
+
+    /**
+     * @notice Event indicating a change of master copy address.
+     * @param singleton New master copy address
+     */
+    event ChangedMasterCopy(address singleton);
+
+    /**
+     * @dev Private function to check if a guard is supported.
+     */
+    function checkGuard() private view {
+        address guard = getGuard();
+
+        if (guard != address(0)) {
+            require(Guard(guard).supportsInterface(type(Guard).interfaceId), "GS300");
+        }
+    }
+
+    /**
+     * @notice Migrate to Safe 1.5.0 Singleton (L1) at `SAFE_150_SINGLETON`
+     * @dev This function should only be called via a delegatecall to perform the upgrade.
+     */
+    function migrateSingleton() public {
+        require(address(this) != MIGRATION_SINGLETON, "Migration should only be called via delegatecall");
+
+        checkGuard();
+
+        singleton = SAFE_150_SINGLETON;
+        emit ChangedMasterCopy(singleton);
+    }
+
+    /**
+     * @notice Migrate and set the fallback handler to Safe 1.5.0 Compatibility Fallback Handler.
+     */
+    function migrateWithFallbackHandler() public {
+        migrateSingleton();
+
+        ISafe(address(this)).setFallbackHandler(SAFE_150_FALLBACK_HANDLER);
+    }
+
+    /**
+     * @notice Migrate and set the guard to the specified address.
+     * @param guard The address of the new guard contract.
+     */
+    function migrateWithSetGuard(address guard) public {
+        require(address(this) != MIGRATION_SINGLETON, "Migration should only be called via delegatecall");
+
+        singleton = SAFE_150_SINGLETON;
```

Comment:
Should we add a require that current value of the `singleton` is something valid (i.e. points to a contract). Else you might change the guard without actually updating the singleton, as a different proxy is used.

Reply:
Added